Search 미션 제출 [0301 내용추가]

seungdeng17 opened the pull request:
데모 : https://seungdeng17.github.io/amazon-DEMO/

- **수정**  
DOM-Templating 반복문 사용을 reduce로 바꾸었습니다.
응답 데이터를 받을 때 fetch api를 해당 데이터를 사용하는 class 메서드로 만들었습니다. (아니면 데이터 통신만 하는 class를 만들어야 할까요?)

- **search 미션 구현2**
방향키 위/아래 입력 시 이동, 엔터키 입력 시 input창에 추가
firebase + express 데이터 통신
최근 검색 결과 노출
webpack, babel 등 적용

#

이전 PR
> - **수정** - fetch 미션에서 피드백 받은 내용을 수정했습니다.
> destructuring 활용 -> Slide 옵션값을 변수화
> 옵션값 대문자 -> 소문자
> if문 긴 조건식 함수로 분리
> NavCard 이벤트 등록 event delegation 방식으로 변경
> 
> - **search 미션 구현1**
> 영단어 1000개에 대한 자동완성결과 노출 (최대 10개까지 응답)
> 300ms 지연 후 화면에 추가
> 입력과 일치하는 단어 하이라이트
> 
> - **추가 할 것**
> 방향키 위/아래 입력 시 이동, 엔터키 입력 시 input창에 추가
> 최근 검색 결과 노출

seungdeng17 committed:
refector. PR 받은 내용 1차 리팩토링

OPTION_DATA -> options 이름 바꾸고 destructuring 문법을 활용해서 변수화 시킴
대문자 옵션값을 소문자로 수정함

seungdeng17 committed:
refector. PR 받은 내용 2차 리팩토링

- Slide class에서 if문 조건식 너무 긴 것 메서드로 분리
- NavCard class 각 li에 이벤트 발생하는 부분 event delegation으로 수정

seungdeng17 committed:
add & update. 자동완성 기능 구현을 위한 베이스 작업

- express 서버 wordSearch 라우터 추가
- 클라이언트에서 body 정보를 전달 받기 위해 body-parser 설치
- input에 값 입력시 300ms 지연 후 서버에 데이터 요청하게 함

seungdeng17 committed:
modify. 클라이언트에서 요청한 글자로 시작하는 단어 10개를 찾으면 즉시 응답하게 수정함

seungdeng17 committed:
add. 검색 동작 위한 검색 UI html + css 추가

seungdeng17 committed:
add. 자동완성 결과 출력 위한 UI html + css 추가

seungdeng17 committed:
add. SearchBar class 작성

seungdeng17 committed:
add. SearchList class 작성

seungdeng17 committed:
feat. search input 창에 입력한 내용과 동일한 부분 하이라이트 처리

seungdeng17 committed:
style. 슬라이드에 폰트 적용, 검색시 검색창을 제외한 전체 화면에 블라인드

seungdeng17 committed:
move & refector. 코드 리뷰 받은 내용 리팩토링

- main 함수 간략화
- js 파일 디렉토리 별로 묶음
- evnet 인자 받을 때 destructuring 문법 활용

seungdeng17 committed:
modify. [ main.js ] -> local storage에 저장되는 slide data key값 변경, 실행순서를 고려해서 함수 위치 조정

crongro commented on index.html:
```diff
@@ -5,26 +5,55 @@
+    <link rel="stylesheet" type="text/css" href="./src/css/reset.css">
+    <link rel="stylesheet" type="text/css" href="./src/css/all.css">
+    <link rel="stylesheet" type="text/css" href="./src/css/slideStyle.css">
+    <link rel="stylesheet" type="text/css" href="./src/css/searchStyle.css">
```
실제 배포에서는 css를 http request를 줄이는 차원에서 합쳐서 배포합니다.
merge한다고 하죠.

seungdeng17 replied:
방학기간을 이용해서 webpack을 공부하도록 하겠습니다!

crongro commented on server/app.js:
```diff
@@ -1,14 +1,20 @@
 app.use(express.static('public'));
+app.use(bodyParser.json());
+app.use(bodyParser.urlencoded({ extended: false }));
+app.use(bodyParser.raw());
```
시간되면 app.js내의 서버설정들이, 
각각 의미하는 것들이 어떤것인지 훑어보고 넘어가면 좋겠네요.

crongro commented on server/public/json/localData.json:
```diff
@@ -1,63 +1,63 @@
-                "출생 : 1988년 5월 12일, 브라질",
-                "신체 : 174cm, 80kg",
-                "소속 : 레알 마드리드 (DF 수비수, 12)"
+                "1988. 5. 12, Brazil",
```
오 영어버전~

crongro commented on server/routes/wordSearch.js:
```diff
@@ -0,0 +1,19 @@
+
+router.post('/', function (req, res, next) {
+    const input = req.body;
+    console.log(input);
```
PR 보낼때 console 지우기!
디버깅은  breakpoint 활용!

crongro commented on server/routes/wordSearch.js:
```diff
@@ -0,0 +1,19 @@
+router.post('/', function (req, res, next) {
+    const input = req.body;
+    console.log(input);
+    const result = [];
```
result보다는 좀더 구체적인 이름으로~

crongro commented on server/routes/wordSearch.js:
```diff
@@ -0,0 +1,19 @@
+    console.log(input);
+    const result = [];
+    const maximum = 10;
+    for (let i = 0; i < wordData.words.length; i++) {
```
for문을 없애고 map을 한번 써볼래요?

seungdeng17 replied:
서버의 부하를 줄이기 위해 클라이언트로 보낼 데이터의 양이 maximum이 됐을 때 반복문이 break 되는 것을 구현하고 싶었습니다.
사용해 본 방식은 reduce, some이 있었지만 원하는 대로 되지 않았어요..ㅜ 다음엔 추천해주신 map을 적용해보도록 하겠습니다.

crongro commented on src/js/lib/util.js:
```diff
@@ -0,0 +1,15 @@
+export function taek$(target) {
```
util 좋아요. 자꾸 늘려보세요.

crongro commented on src/js/main.js:
```diff
@@ -0,0 +1,48 @@
+    startSeachBarSevice();
+}
+
+main();
```
main은 즉시실행함수로 표현해도 될 듯.

seungdeng17 replied:
오늘 강의를 통해 즉시실행함수를 알게되었습니다. 다음 리팩토링에 적용해보도록 하겠습니다!

crongro commented on src/js/main.js:
```diff
@@ -0,0 +1,48 @@
+    if (data) {
+        startSlideService(JSON.parse(data));
+    } else {
+        fetch("http://localhost:8080/json/localData.json")
```
URL을 적어도 함수의 인자로 받는 게 어떨까요? 이런게 하드코딩되어 있는건 안좋습니다.
url은 변경될 수 있거든요. 개발단계의 URL과 실제 배포시의 URL이 다르기도 하고요.

crongro commented on src/js/search/searchBar.js:
```diff
@@ -0,0 +1,58 @@
+}
+
+SearchBar.prototype = {
+    constructor: SearchBar,
```
👍

crongro commented on src/js/search/searchBar.js:
```diff
@@ -0,0 +1,58 @@
+            body: input
+        })
+            .then(rep => rep.json())
+            .then(json => {
```
여기 then함수 인자로 들어가는 함수를 분리해보세요.

crongro commented on src/js/search/searchBar.js:
```diff
@@ -0,0 +1,58 @@
+            this.searchList.searchResultOff();
+            classRemove(this.serachBlind, "on");
+        } else {
+            this.timer = setTimeout(this.wordDataRequest.bind(this), 300, input);
```
오 setTimeout 세번째인자!!
브라우저 호환성도 한번 확인해보세요.

crongro commented on src/js/search/searchBar.js:
```diff
@@ -0,0 +1,58 @@
+    wordDataRequest(input) {
+        fetch("http://localhost:8080/wordSearch", {
+            method: 'POST',
+            body: input
```
데이터가 잘 전송되는지 개발자도구의 네트워크 탭을 확인해보시죠!

crongro commented on src/js/search/searchList.js:
```diff
@@ -0,0 +1,29 @@
+
+    renderList(data, input = "") {
+        let temp = "";
+        data.forEach(word => {
```
reduce로 고쳐봅시다. 
좀더 간결하게요.

seungdeng17 replied:
reduce를 쓰면 지역변수를 쓰지 않아도 될 것 같아요 감사합니다!

crongro reviewed (changes requested):
전체 왁꾸(?)는 좋습니다.

리뷰 남긴건 참고해서 수정하고, 기능을 추가해서 다시 push!

seungdeng17 committed:
refector. wordSearch for문 -> reduce, main.js 간소화, slide data 가져오는 부분 DataRender class에 넣음, 오타 수정

seungdeng17 committed:
refector. 피드백 받은 내용 수정

- main.js -> main 함수 즉시실행함수로 변경
- searchBar.js -> 콜백함수 메서드로 분리 (searchDataDeliver)
- searchList -> 반복문으로 DOM-Templating 하는 부분 reduce로 변경
- dataRender.js -> 반복문으로 DOM-Templating 하는 부분 reduce로 변경, dummy node 생성하는 메서드 재사용 가능하게 분리

seungdeng17 committed:
fix. slide data fetch 받는 위치를 DataRender class로 옮기면서 생긴 문제 수정

데이터를 받아 오기전에  slide, navCard 세팅 메서드들이 호출돼서 async, await으로 대기시킴

seungdeng17 committed:
add. 키보드 방향키 입력 -> 지동완성 하이라이트, 엔터 입력 -> searchInput value값 변경 기능 추가

seungdeng17 committed:
refector. searchList 방향키 입력시 발생하는 이벤트 중복되는 코드 제거

seungdeng17 committed:
modify. firebase 호스팅 가능한 버전으로 수정

seungdeng17 committed:
modify. slide data 통신 방식 post -> get으로 수정

seungdeng17 committed:
add. submit 이벤트시 사용자가 입력한 검색값 cache, 검색창에 포커스가 가면, 최근 검색한 결과가 최대 5개까지 노출되게하는 기능 추가

seungdeng17 committed:
modify. options.js에 searchDelayTime property 추가 -> 매직넘버 수정

seungdeng17 committed:
modify. 처음 보여질 slide item을 랜덤하게 지정하는 코드 -> Slide class에서 초기화 해주는 것으로 수정

seungdeng17 committed:
update. webpack 빌드 babel, css-loader, file-loader 적용 -> js, css, 기타 파일 번들링

seungdeng17 committed:
modify. publicPath 경로 수정

crongro commented on server/app.js:
```diff
@@ -3,6 +3,7 @@ var cors = require('cors');
 var bodyParser = require('body-parser');
 var app = express();
 
+var slideDataRouter = require('./routes/slideData');
```
node코드도 es6문법을 사용하는게 좋겠네요.

crongro commented on server/routes/wordSearch.js:
```diff
@@ -1,19 +1,16 @@
-    }
-    res.json(result);
+    const responseData = wordData.words.reduce((acc, word) => {
+        if (acc.length >= maximum || !word.startsWith(input)) return acc;
```
maximum 보다는 maxCount라던가 뭔이 이름을 제대로~

crongro commented on src/js/common/url.js:
```diff
@@ -0,0 +1,12 @@
+        SLIDE_DATA_API: "http://localhost:8080/slideData",
+        SEARCH_DATA_API: "http://localhost:8080/wordSearch",
+    },
+    PROD: {
```
좋네요.
보통 production 을 잘 줄이지는 않아요.

crongro commented on src/js/search/options.js:
```diff
@@ -0,0 +1,7 @@
+const options = {
```
options내용은 분리하는 것보다 searchBar에 두고 의도를 드러내는게 좀더 좋다고 생각되네요.

crongro commented on src/js/search/searchList.js:
```diff
@@ -1,28 +1,119 @@
+            }
+                break;
+            case "Enter": {
+                if (this.curList !== null) {
```
이만큼정도도 함수로 분리할 수도 있긴하죠.
```javascript
if (this.curList !== null) {
  evt.preventDefault();
  classRemove(taek$(".search-blind"), "on");
  this.searchListOff();
  this.listInit();
  this.searchInput.value = this.curList.innerText;
```

crongro commented on src/js/search/searchList.js:
```diff
@@ -1,28 +1,119 @@
+        this.prevList = this.curList.previousElementSibling;
+    },
+
+    curListStateCheck(isKeyDown) {
```
checkCurrentListState 라고 하는게 명확할 듯.

crongro commented on src/js/search/searchList.js:
```diff
@@ -1,28 +1,119 @@
+    curListStateCheck(isKeyDown) {
+        if (this.curList === this.searchList.firstElementChild && !isKeyDown) return true
+        if (this.curList === this.searchList.lastElementChild && isKeyDown) return true
+        if (this.curList === null) return true
```
로직이 맞다고 하더라도, 
null인데 true를 반환하는 것이 어색해보이네요. 음

crongro commented on src/js/slide/dataRender.js:
```diff
@@ -1,77 +1,81 @@
+        if (lacalStorageData) {
+            this.data = JSON.parse(lacalStorageData);
+        } else {
+            return new Promise(resolve => {
```
promise를 반환시켰군요. 
나중에 async함수를 사용하면 promise객체로 자동반환 되긴합니다.

crongro reviewed (approved):
수고하셨네요. 
리뷰 남기고 머지승인합니다~
